ast/impl/ast_printer.cpp

```diff
@@ -0,0 +1,288 @@
+
+    m_depth++;
+
+    if (node->actualType != nullptr) {
+        node->actualType->accept(*this);
+    }
+
+    m_depth--;
+}
+
+void AstPrinter::visit(PrimitiveType*) {
+    fmt::print("{:|>{}}- [PrimitiveType]> unknown\n", "", m_depth);
+}
+
+void AstPrinter::visit(IntegerType*) {
+    fmt::print("{:|>{}}- [IntegerType]\n", "", m_depth);
+}
+
+void AstPrinter::visit(RealType*) {
+    fmt::print("{:|>{}}- [RealType]\n", "", m_depth);
+}
+
+void AstPrinter::visit(BooleanType*) {
+    fmt::print("{:|>{}}- [BooleanType]\n", "", m_depth);
+}
+
+void AstPrinter::visit(ArrayType* node) {
+    fmt::print("{:|>{}}- [ArrayType]>\n", "", m_depth);
+
+    m_depth++;
+
+    if (node->length != nullptr) {
+        node->length->accept(*this);
+    }
+
+    if (node->elementType != nullptr) {
+        node->elementType->accept(*this);
+    }
+
+    m_depth--;
+}
+
+void AstPrinter::visit(RecordType* node) {
+    fmt::print("{:|>{}}- [RecordType]>\n", "", m_depth);
+
+    m_depth++;
+
+    for (auto field : node->fields) {
+        field->accept(*this);
+    }
+
+    m_depth--;
+}
+
+void AstPrinter::visit(VariableDecl* node) {
+    fmt::print("{:|>{}}- [VariableDecl]> {}\n", "", m_depth, node->name);
+
+    m_depth++;
+
+    if (node->type != nullptr) {
+        node->type->accept(*this);
+    }
+
+    if (node->initialValue != nullptr) {
+        node->initialValue->accept(*this);
+    }
+
+    m_depth--;
+}
+
+void AstPrinter::visit(TypeDecl* node) {
+    fmt::print("{:|>{}}- [TypeDecl]> {}\n", "", m_depth, node->name);
+
+    m_depth++;
+
+    if (node->type != nullptr) {
+        node->type->accept(*this);
+    }
+
+    m_depth--;
+}
+
+void AstPrinter::visit(Body* node) {
+    fmt::print("{:|>{}}- [Body]>\n", "", m_depth);
+
+    m_depth++;
+
+    for (auto type : node->types) {
+        type->accept(*this);
+    }
+
+    for (auto variableDecl : node->variables) {
+        variableDecl->accept(*this);
+    }
+
+    for (auto statement : node->statements) {
+        statement->accept(*this);
+    }
+
+    m_depth--;
+}
+
+void AstPrinter::visit(Statement*) {
+    fmt::print("{:|>{}}- [Statement]> Stray\n", "", m_depth);
+}
+
+void AstPrinter::visit(ReturnStatement* node) {
+    fmt::print("{:|>{}}- [Return]>\n", "", m_depth);
+    
+    m_depth++;
+
+    if (node->expression != nullptr) {
+        node->expression->accept(*this);
+    }
+
+    m_depth--;
+}
+
+void AstPrinter::visit(Assignment* node) {
+    fmt::print("{:|>{}}- [Assignment]>\n", "", m_depth);
+
+    m_depth++;
+
+    node->lhs->accept(*this);
+    node->rhs->accept(*this);
+
+    m_depth--;
+}
+
+void AstPrinter::visit(WhileLoop* node) {
+    fmt::print("{:|>{}}- [WhileLoop]>\n", "", m_depth);
+
+    m_depth++;
+
+    node->condition->accept(*this);
+    node->body->accept(*this);
+
+    m_depth--;
+}
+
+void AstPrinter::visit(ForLoop* node) {
+    fmt::print("{:|>{}}- [ForLoop]>\n", "", m_depth);
+
+    m_depth++;
+
+    // node->loopVar->accept(*this);
```

Comment:
`loopVar`'s type has already been changed from `Token` to `VariableDecl`, so this line can be uncommented